test/markup_renderer.go

```diff
@@ -0,0 +1,497 @@
+	attrs := map[string]*string{}
+	r.setSizeAttr(attrs, "size", c.Size())
+	if tc, ok := c.(WindowlessCanvas); ok {
+		r.setBoolAttr(attrs, "padded", tc.Padded())
+	}
+	r.writeTag("canvas", false, attrs)
+	r.w.WriteRune('\n')
+	r.indentation++
+	r.writeTag("content", false, nil)
+	r.w.WriteRune('\n')
+	r.indentation++
+	driver.WalkVisibleObjectTree(c.Content(), r.writeCanvasObject, r.writeCloseCanvasObject)
+	r.indentation--
+	r.writeIndent()
+	r.writeCloseTag("content")
+	for _, o := range c.Overlays().List() {
+		r.writeTag("overlay", false, nil)
+		r.w.WriteRune('\n')
+		r.indentation++
+		driver.WalkVisibleObjectTree(o, r.writeCanvasObject, r.writeCloseCanvasObject)
+		r.indentation--
+		r.writeIndent()
+		r.writeCloseTag("overlay")
+	}
+	r.indentation--
+	r.writeIndent()
+	r.writeCloseTag("canvas")
+}
+
+func (r *markupRenderer) writeCanvasObject(obj fyne.CanvasObject, _, _ fyne.Position, _ fyne.Size) bool {
+	attrs := map[string]*string{}
+	r.setPosAttr(attrs, "pos", obj.Position())
+	r.setSizeAttr(attrs, "size", obj.Size())
+	switch o := obj.(type) {
+	case *canvas.Circle:
+		r.writeCircle(o, attrs)
+	case *canvas.Image:
+		r.writeImage(o, attrs)
+	case *canvas.Line:
+		r.writeLine(o, attrs)
+	case *canvas.LinearGradient:
+		r.writeLinearGradient(o, attrs)
+	case *canvas.RadialGradient:
+		r.writeRadialGradient(o, attrs)
+	case *canvas.Raster:
+		r.writeRaster(o, attrs)
+	case *canvas.Rectangle:
+		r.writeRectangle(o, attrs)
+	case *canvas.Text:
+		r.writeText(o, attrs)
+	case *fyne.Container:
+		r.writeContainer(o, attrs)
+	case fyne.Widget:
+		r.writeWidget(o, attrs)
+	case *layout.Spacer:
+		r.writeSpacer(o, attrs)
+	default:
+		panic(fmt.Sprint("please add support for", reflect.TypeOf(o)))
+	}
+
+	return false
+}
+
+func (r *markupRenderer) writeCircle(c *canvas.Circle, attrs map[string]*string) {
+	r.setColorAttr(attrs, "fillColor", c.FillColor)
+	r.setColorAttr(attrs, "strokeColor", c.StrokeColor)
+	r.setFloatAttr(attrs, "strokeWidth", float64(c.StrokeWidth))
+	r.writeTag("circle", true, attrs)
+}
+
+func (r *markupRenderer) writeCloseCanvasObject(o, _ fyne.CanvasObject) {
+	switch o.(type) {
+	case *fyne.Container:
+		r.indentation--
+		r.writeIndent()
+		r.writeCloseTag("container")
+	case fyne.Widget:
+		r.indentation--
+		r.writeIndent()
+		r.writeCloseTag("widget")
+	}
+}
+
+func (r *markupRenderer) writeCloseTag(name string) {
+	r.w.WriteString("</")
+	r.w.WriteString(name)
+	r.w.WriteString(">\n")
+}
+
+func (r *markupRenderer) writeContainer(c *fyne.Container, attrs map[string]*string) {
+	r.writeTag("container", false, attrs)
+	r.w.WriteRune('\n')
+	r.indentation++
+}
+
+func (r *markupRenderer) writeIndent() {
+	for i := 0; i < r.indentation; i++ {
+		r.w.WriteRune('\t')
+	}
+}
+
+func (r *markupRenderer) writeImage(i *canvas.Image, attrs map[string]*string) {
+	r.setStringAttr(attrs, "file", i.File)
+	r.setResourceAttr(attrs, "rsc", i.Resource)
+	r.setBoolAttr(attrs, "img", i.Image != nil)
```

Comment:
Ditto; `image`

Reply:
I think that img is a pretty reasonable abbreviation. It only removes 2 letters and "file" is 4 so I don't have strong feelings either way.

Reply:
The vote has clearly favoured the abbreviations (no one voted for full names).